### DISCUSSION TITLE
Strategies for managing performance for many bindings

### DISCUSSION THREAD

**Discussion by jeffective** [zauberzeug/nicegui#2876] [Q&A]
### Question

I have a machine control application where the machine is emitting roughly 2000 channels of data at 500 Hz to a redis instance using redis streams.

I've defined a sort of "backend" object that polls redis every 250 ms to maintain a mirror of the 2000 channels as a dictionary that I can bind to in my nicegui application to display live data for the user.

I've found that I cannot simply show 2000 ui.input or ui.label elements to display these things live for the user as the binding propagation will take too long.

What are some ways y'all have implemented to manage a situation like this? Here are some of the things I have thought of:

1.  Using ui.tabs to split up the data into multiple views. (does this actually work? will non-displayed elements not update?
2. Repackaging the data into a large markdown table and updating a single massive ui.markdown element at once every 250 ms (this is very CPU intensive and limiting for a machine control application that requires buttons etc.)
3. using a multipage application (would prefer to do a single page application) in similar manner to the ui.tabs idea.
4. doing some sort of lazy loading behavior when elements are not on screen (is this possible)?

**Comment by jeffective** [comment]
and if anyone is wondering, if you use redis + msgspec the deserialization for all 2000 channels can be on the order of 20 us :)

**Comment by Alyxion** [comment]
Do they regularly nearly all change or just a small subset? In the first case and as I guess your end-user is non the T1000 but a human and that a 2 second update interval is more than enough I would very likely do something like the following:

Pseudo-code:

```py
labels ={}
current_values = {}
total_update_turns = 8
total_update_time = 2.0
elements_per_turn = int(math.ceil(len(data_id_list)/total_update_turns))
cur_update_turn = 0
for cur_id in data_id_list: # data_id list contains all valid ids (or just the indices from 0..1999)
   cur_val = previous_values[cur_id] = your_cache[cur_id]
   labels[cur_id] = ui.label(f"{cur_id}: {cur_val}")
   
def update_data():
   start = cur_update_turn*elements_per_turn   
   ids = data_id_list[start:start+elements_per_turn]
   for cur_id in ids:
       if previous_values[cur_id]!=your_cache[cur_id]:
          cur_val = previous_values[cur_id] = your_cache[cur_id]
          labels[cur_id].set_text(f"{cur_id}: {cur_val}")
   cur_update_turn = (cur_update_turn+1)%total_update_turns
          
ui.timer(total_update_time/total_update_turns, update_data, once=False)
```

This will do 8 turns in 2 seconds, each one comparing 12.5% of the whole content and sending update requests for the affected labels.

On top you can put your elements into separate divs and then something like this:
```py
window.addEventListener('scroll', () => {
    const element = document.querySelector("#yourDidIds");
    console.log(isInViewport(element) ? "Visible" : "Not visible");
});
```

for your divs to inform yourself which of all the elements you need to update at all.
